java/src/main/java/com/google/oak/util/Result.java

```diff
@@ -0,0 +1,148 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.oak.util;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Class {@code Result} is a generic class representing the result of an
+ * operation that could either succeed and produce an instance of {@code R}, or
+ * fail and produce an error of type {@code E}. An instance of {@code Result}
+ * can encapsulate both outcomes.
+ *
+ * <p>The main invariant that instances of this class maintain is that each object always either
+ * contains a non-null success value or a non-null error value.
+ */
+public class Result<R, E> {
+  private final Optional<R> success;
+  private final Optional<E> error;
+
+  /**
+   * Create and return an instance of {@code Result<E, R>} wrapping {@code success}
+   * as the success value.
+   *
+   * @param success nonnull success value
+   * @return a {@code Result} containing a success value
+   */
+  public static <R, E> Result<R, E> success(final R success) {
+    Objects.requireNonNull(success);
```

Comment:
API-wise, with `null` as a valid success value, we'd get into a problem with `success()` and `error()`. If we allow `null` here, we'll have an object for which both `success()` and `error()` would return an empty `Optional`. In this case it is not possible to distinguish between a success or an error case. If `null` values are valid as the result of an operation, then `Optional<R>` must be used, instead of `<R>` as the type of success value. I thought about adding a comment about it, but I am not sure this is something that we'd want to include in the javadoc.